### PR TITLE
feat(android): Enhance how ENTER key is handled in apps

### LINKED ISSUE #7796: feat(android): Allow enter key to send newline instead of submit in messenger apps
### Is your feature request related to a problem? Please describe.

A review comment from the Google Play Store (A Google user November 7, 2019)

> My father uses this app exclusively to write on the Facebook and Facebook messenger apps. There is no return key to make a new paragraph. If you press the enter key for either of these apps it sends the message instead of entering a space or creating a new paragraph. In other keyboard apps he is able to press shift + enter to create a new line or paragraph. If there was a way to do this it would be perfect for him. He only writes using polytonic Greek and is very frustrated since other android keyboards don't have all the accents he needs.

### Describe the solution you'd like

The user would like a way for enter to send a new line instead of submitting text in a messenger app.

(Do the text fields provide [enterkeyhint](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/enterkeyhint)?)

### Describe alternatives you've considered

_No response_

### Related issues

_No response_

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

_No response_

### Operating system

_No response_

### Device

_No response_

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #7796

This PR enhances how the ENTER key is handled as actions in different apps.

In multi-line text (like messaging apps), the enter key inserts a newline (`\n`).

When the IME action is Go / Search / Done, the ENTER key sends the corresponding action. (DONE dismisses the keyboard).
Note: [EditorInfo](https://developer.android.com/reference/android/view/inputmethod/EditorInfo) also has more actions like PREVIOUS and NEXT, but I didn't bother implementing those this time.
(edit)
From standup, I'll add PREVIOUS and NEXT actions.

Also a future PR for KMW is getting the <kbd>ENTER</kbd> key label to change depending on the action

## API Changes
to add to help.keyman.com
* KMManager.setEnterMode
* KMManager.getEnterMode

## User Testing
**Setup** - On an Android device, install the PR build of Keyman for Android. Preferably use a physical device instead of emulator because you'll also need additional Android apps installed:
* Messaging app like Facebook Messenger
* X (twitter)

Setup and enable Keyman as a default system keyboard

* **TEST_KEEP** - Verifies how ENTER key is handled in Keep app
1. Launch Keep
2. With Keyman selected as the system keyboard, type in the Keep app
3. Press the "ENTER" key
4. Verify a newline is inserted in the text

* **TEST_MESSAGING** - Verifies how ENTER key is handled in a multi-line text field (as requested in the original issue)
1. Launch messaging app (like Facebook Messenger)
2. Select a contact to message and start typing a message with Keyman as the system keyboard
3. Press the "ENTER" key
4. Verify a newline is inserted in the text
5. Above the keyboard, the device should have an arrow key to send the message
6. Press the arrow key
7. Verify the message is sent

* **TEST_SEARCH** - Verifies how ENTER key is handled in a search field (as @dyacob requested in the original issue)
* Launch X (Twitter app)
* Click the eyeglass icon to start a search
* Start typing in the search field with Keyman as the system keyboard
* Press the "ENTER" key
* Verify the search is triggered

### CODE REVIEW

**Comment by dinakaranr** [comment]
# Test Results
I tested this issue with the attached "Keyman 18.0.86-alpha-test-12125" build on the Android 14 physical device. Here is my observation.
1. Installed the "Keyman-18.0.88.apk" file and gave all permissions to the application.
2. Checked the "Enable Keyman as system-wide keyboard" and set the keyboard as the default keyboard box on the settings page.
3. Open the Keyman app. Enable the "Predictions" to install the "Dictionary."

* **TEST_KEEP (Passed):** 
1. Launch the Google Keep app.
2. Open the empty note by clicking the "+" button.
3. Enter a sentence using the keyman keyboard.
4. Press the "Enter" key.
5. I observed that an empty line was inserted under the sentence. 

* **TEST_MESSAGING (Failed):** 
1. Launch the Facebook messenger app.
2. Select a person from the contact list.
3. Enter a sentence using the keyman keyboard.
4. Press the "Enter" key.
5. I observed that the empty line was inserted under the sentence. 
6. Click the arrow key.
Observed that the message was sent to the user. but the blank line was removed in the sent message. (Which is inserted a line.)
It is a problem. Please refer to the attached video.
https://github.com/user-attachments/assets/131ee1b7-d6a1-48a9-a955-941a0973f6a2

* **TEST_SEARCH (Passed):** 
1. Launch the Twitter app.
2. Open the search box by clicking the eyeglass icon.
3. Enter a word in the search field.
4. Press the "Enter" key.
Observed that the search result appears after pressing the "Enter" key. It works well. Thank you.

**Comment by darcywong00** [comment]
@dinakaranr - I've made a code change to advance the cursor on newline.
However, on my default Gboard keyboard, entering multiple blank newlines in the messaging app ends up with the newlines also getting truncated when the text is sent.

@keymanapp-test-bot retest TEST_MESSAGING

**Comment by dinakaranr** [comment]
# Test Results

* **TEST_MESSAGING (Passed):**
1. Launch the Facebook messenger app.
2. Select a person from the contact list.
3. Enter a sentence using the keyman keyboard.
4. Press the "Enter" key.
5. I observed that the empty line was inserted under the sentence.
6. The mouse cursor moved into a new blank line. 
7. Click the arrow key.
Observed that the message was sent to the user. but the blank line was removed from the sent message. This works as expected behavior. 
Thank you Darcy for the clarification.

File: android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java [hunk 1]
```diff
@@ -153,7 +154,53 @@ public void run() {
...
+                break;
+
+              // Default ENTER action
+              default:
+                keyDownUp(KeyEvent.KEYCODE_ENTER, 0);
+            }
+          } else {
+            // In-app keyboard uses default ENTER action
+            keyDownUp(KeyEvent.KEYCODE_ENTER, 0);
+          }
```
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboardJSHandler.java:203]: This seems like it is an opportunity for a future feature?

**Review by mcdurdin — APPROVED** [review]
LGTM. We need a new issue to make the Enter mode visible on the <kbd>Enter</kbd> key on the touch keyboard.

Can <kbd>Shift</kbd> layer change the Enter mode on other keyboard apps?

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.92-alpha](https://keyman.com/downloads/releases/alpha/18.0.92)

**Pull request merged**